Implemented @Entity Employee, EmployeeDao implementation is now via J…

RomanZubariev opened the pull request:
…paRepository.

RomanZubariev1990 committed:
Implemented @Entity Employee, EmployeeDao implementation is now via JpaRepository.

RomanZubariev1990 committed:
Implemented global exception handling in one class

HalinaPichukha commented on src/main/java/com/mastery/java/task/config/AppConfiguration.java:
```diff
@@ -9,7 +9,7 @@
 
   @Bean
-  SimpleJdbcInsert simpleJdbcInsert(DataSource dataSource){
+  SimpleJdbcInsert simpleJdbcInsert(DataSource dataSource) {
```
А где у тебя используется SimpleJdbcInsert  бин?

RomanZubariev replied:
Не используется. Убрал.

HalinaPichukha commented on src/main/java/com/mastery/java/task/dao/EmployeeDaoImpl.java:
```diff
@@ -1,79 +1,51 @@
-  private final JdbcTemplate jdbcTemplate;
-  private final SimpleJdbcInsert simpleJdbcInsert;
-  private final BeanPropertyRowMapper<Employee> beanPropertyRowMapper;
+  private EmployeeRepository employeeRepository;
```
Забыл про ключнвое слово)

RomanZubariev replied:
Изменил на final.

HalinaPichukha commented on src/main/java/com/mastery/java/task/dao/EmployeeDaoImpl.java:
```diff
@@ -1,79 +1,51 @@
+    return employeeRepository.findById(id).get();
   }
 
   //read the whole table as a List of Employee
```
Комментарии бывают полезны очень редко) Описание очевидных вещей загрязняет код. Вместо комментариев методы должны быть понятно названы и выполнять понятную раюоту. Для описания public методов есть java docs (можешь попробовать, кстати. Обычно на проектах описание public методов сервисов java доками - маст хэв)

RomanZubariev replied:
Убрал.

HalinaPichukha commented on src/main/java/com/mastery/java/task/dao/EmployeeDaoImpl.java:
```diff
@@ -1,79 +1,51 @@
-    return jdbcTemplate.queryForObject(GET_BY_ID_SQL, beanPropertyRowMapper, id);
+  @Override
+  public Employee getById(Long id) {
+    return employeeRepository.findById(id).get();
```
У тебя здесь возвращается объект типа Optional. И перед тем, как делать гет, нужно проверить, что он что-то содержит. Если присмотришься к подсказкам идеи, то заметишь, что она пытается помочь)

RomanZubariev replied:
заменил на orElseThrow(). В случае пустого  optional выкидывается эксепшн. Я их планировал централизовано чуть позже обработать.

HalinaPichukha commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -16,15 +16,15 @@ public class EmployeeServiceImpl implements EmployeeService {
-    } catch (EmptyResultDataAccessException e) {
-      e.printStackTrace();
+    } catch (NoSuchElementException e){
       throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Employee not found");
```
"Employee not found" в классе встречается 3 раза

RomanZubariev replied:
Сделал разные сообщения чтобы не перепутать их потом.

HalinaPichukha replied:
Если используются одинаковые значения, то их нужно выносить в константы)

RomanZubariev replied:
Имеет ли смысл выносить в константы сообщения эксепшнов?

RomanZubariev replied:
Сообщения эксепшнов в принципе выносят в константы?

HalinaPichukha commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -42,17 +42,21 @@ public Employee save(Employee employee) {
+    } catch (EmptyResultDataAccessException e) {
+      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Employee not found");
+    }
+
```
форматирование

RomanZubariev1990 committed:
Removed unused dependencies, beans and comments

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/dao/EmployeeDaoImpl.java:
```diff
@@ -1,79 +1,47 @@
-  private final JdbcTemplate jdbcTemplate;
-  private final SimpleJdbcInsert simpleJdbcInsert;
-  private final BeanPropertyRowMapper<Employee> beanPropertyRowMapper;
+  private final EmployeeRepository employeeRepository;
```
Чем отличается DAO от Repository? Зачем нам дополнительный слой в виде DAO здесь?

pavel-tsikhanovich replied:
Можно закрыть, отвечено в скайпе

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/dao/EmployeeRepository.java:
```diff
@@ -0,0 +1,8 @@
+import com.mastery.java.task.dto.Employee;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface EmployeeRepository extends JpaRepository<Employee, Long> {
```
Какие ещё виды репозиториев есть в в Spring Data? Какой самый простой для нашего набора операций?

RomanZubariev replied:
Не знаю насчет классификации. JpaRepository наследует CrudRepository, которого должно хватать на простые CRUD операции (как у нас). JpaRepository также наследует PagingAndSortingRepository который позволяет работать с порциями данных (страницами из объектов) и еще дает для сортировки методы, плюс добавляет методы для удаления групп объектов (batch) одним запросом.

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/rest/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+  @ExceptionHandler(EmptyResultDataAccessException.class)
+  public final ResponseEntity handleEmptyResultDataAccessException(
```
Почему тут все методы помечены как final? Что нам это дает?

RomanZubariev replied:
Я не собирался запрещать оверрайдить методы (как впрочем и наследовать класс). Убрал final.

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/rest/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+    return new ResponseEntity(HttpStatus.NOT_FOUND);
+  }
+
+  @ExceptionHandler(NoSuchElementException.class)
```
`EmptyResultDataAccessException` и `NoSuchElementException` в каком-то смысле очень похожи, а логика у тебя тем более одинаковая, может можно как-то их объединить?

pavel-tsikhanovich replied:
Можно закрыть,

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -16,36 +14,30 @@ public class EmployeeServiceImpl implements EmployeeService {
+    if (employeeDao.update(employee) != null) {
+      return employee;
+    } else {
       throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Employee not found");
```
Раз тут уже Exception handler есть, то вот этот эксепшн очень похож на другие твои, может стоит переделать или использовать какой-либо 1 (если ты сам его используешь, а не API выбрасывает)

pavel-tsikhanovich replied:
Можно закрыть,

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -16,36 +14,30 @@ public class EmployeeServiceImpl implements EmployeeService {
-      return employeeDao.update(employee);
-    } catch (EmptyResultDataAccessException e) {
-      e.printStackTrace();
+    if (employeeDao.update(employee) != null) {
+      return employee;
+    } else {
```
Посмотри, пожалуйста, на API класса `Optional`, может красивее тут сделаешь

pavel-tsikhanovich replied:
Можешь это закрыть

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/dto/Employee.java:
```diff
@@ -1,15 +1,28 @@
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
   private Long employeeId;
+
   private String firstName;
```
А как нам быть, если имя поля не совпадает с именем столбца в таблицу БД? Ещё что мы можем сделать, если не хотим, если какое-нибудь поле сохранялось в БД? 
Это просто теория, хочу увидеть ответы)

RomanZubariev replied:
Можно задать свойства столбца аннотацией @Column(name = "not_first_name"). Опциями  insertable и updatable можно запретить конкретно sql с insert и update.

pavel-tsikhanovich replied:
A чтобы оно вообще не менеджилось ничем, а хранилось только в памяти? В джаве есть похожее ключевое слово

RomanZubariev replied:
@ Transient

RomanZubariev1990 committed:
Removed DAO classes as their job is done by EmployeeRepository bean now

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -1,58 +1,55 @@
 
-  private final EmployeeDao employeeDao;
+  private final CrudRepository<Employee, Long> employeeRepository;
+  private EmptyResultDataAccessException notFoundException;
```
А вот так лучше уже не делать, просто выбрасывай новые исключения там где надо)

RomanZubariev replied:
Да, пожалуй :)

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -1,58 +1,55 @@
   @Override
   public List<Employee> getAll() {
-    return employeeDao.getAll();
+    List<Employee> employeeList = new ArrayList<>();
+    employeeRepository.findAll().forEach(employeeList::add);
+    return employeeList;
```
Почему ты так решил сделать?

RomanZubariev replied:
findAll возвращает Iterable. Перевел его в List. List это уже Collection, с ней в перспективе удобней работать.

pavel-tsikhanovich replied:
В перспективе да, но сейчас это выглядит как ненужное действие для меня. Возможно, используя стримы можно было бы это сделать проще

RomanZubariev replied:
Оверрайднул findAll() в интерфейсе репозитория чтобы он возвращал List.

pavel-tsikhanovich replied:
Не то, чего я ожидал. Давай вернем все обратно, просто вместо вот этого вот когда наверху ты из Iterable сделаешь List с помощью Stream, хорошо? Ну или верни обратно Iterable и просто передавай ещё вверх, тоже окей)

RomanZubariev replied:
Там Spring boot просто дает бин с нужной имплементацией. 
Сделал конвертацию как раньше, но без итераций (через Streamable).

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -1,58 +1,55 @@
   @Override
   public Employee save(Employee employee) {
-    return employeeDao.save(employee);
+    employee.setEmployeeId(null);
```
Почему решил явно сетать `null` тут? Может это все-таки ошибка, когда собираются сохранить нового Employee с ID?

RomanZubariev replied:
Поставил условие на id. Теперь сохранять Employee с id это ошибка.

pavel-tsikhanovich reviewed (commented):
Давай попробуем уже делать хоть какие-нибудь тесты, можно начать с Service слоя

RomanZubariev1990 committed:
Minor changes to exception handling. POST employee with a non-null id will throw an IllegalArgumentException now

RomanZubariev1990 committed:
Minor formatting

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -34,16 +33,20 @@ public List<Employee> getAll() {
   public Employee save(Employee employee) {
-    employee.setEmployeeId(null);
-    return employeeRepository.save(employee);
+    if (employee.getEmployeeId() == null) {
+      return employeeRepository.save(employee);
+    } else {
+      throw new IllegalArgumentException("Employee id should be null.");
+    }
```
Давай попробуем `Optional` здесь

RomanZubariev replied:
Использовал Optional вместо if.

pavel-tsikhanovich commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -34,16 +33,20 @@ public List<Employee> getAll() {
     } else {
-      throw notFoundException;
+      throw new EmptyResultDataAccessException(
+          "Cannot update employee: id " + employee.getEmployeeId() + " not found", 1);
```
И здесь поиграйся с `Optional`

RomanZubariev replied:
Поигрался.

RomanZubariev1990 committed:
added loggers dependencies to pom

RomanZubariev1990 committed:
Toyed with Optionals to replace conditional statements for readability improvement

RomanZubariev1990 committed:
Overrided the findAll() method in EmployeeRepository to return List instead of Iterable and removed the conversion in EmployeeServiceImpl

RomanZubariev1990 committed:
Added service layer tests.

RomanZubariev1990 committed:
EmployeeServiceImpl.getAll() is once again converting Iterable to List (via Streamable)

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+import org.springframework.dao.EmptyResultDataAccessException;
+
+@ExtendWith(MockitoExtension.class)
 public class EmployeeServiceTest {
```
Модификатор не нужен - "JUnit5 test classes and methods should have default package visibility"
Это можно найти в документации JUnit5

RomanZubariev replied:
убрал.

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+  }
+
+  @Test
+  public void getEmployeeTest() {
```
Модификаторы для тестов тоже не нужны.

RomanZubariev replied:
Убрал.

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+
+  @Test
+  public void getNullEmployeeTest() {
+    when(employeeRepository.findById(100L)).thenReturn(Optional.ofNullable(null));
```
Есть другой метод для получения пустого объекта Optional без лишних проверок

RomanZubariev replied:
.empty()

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+  public void getNullEmployeeTest() {
+    when(employeeRepository.findById(100L)).thenReturn(Optional.ofNullable(null));
+    Assertions.assertThrows(EmptyResultDataAccessException.class,
+        () -> employeeServiceImpl.getById(100L));
```
Мы можем проверить не только тип исключение, но и конкретный message

RomanZubariev replied:
Отдельным assertTrue проверил месседж и id в месседже.

HalinaPichukha replied:
getNullEmployeeTest() падает ().
+ Method invocation 'contains' may produce 'NullPointerException'
+ зачем проверять на contains, если можно проверить полное сообщение и убедиться, что мы действительно увидим id в нём?

RomanZubariev replied:
Проверил на равенство сообщения эксепшна ожидаемому (вместе с id).

HalinaPichukha commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -1,58 +1,55 @@
   @Override
   public Employee save(Employee employee) {
-    return employeeDao.save(employee);
+    Optional.ofNullable(employee.getEmployeeId()).ifPresent((id) -> {
```
В лямбдах в скобки обычно берутся выражение, так что id лучше без скобок указывать

RomanZubariev replied:
убрал

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+  @Test
+  public void getEmployeeTest() {
+    when(employeeRepository.findById(80L)).thenReturn(Optional.ofNullable(testEmployee));
+    Assertions.assertEquals(testEmployee, employeeServiceImpl.getById(80L));
```
Если метод в сервисе делает простую работу, это не значит, что нечего тестить)) 
Чтобы немного расширить знания по тестированию, можешь проверить, что конкретный метод у конкретного сервиса с ожидаемыми параметрами был вызван ожидаемое колличество раз.

RomanZubariev replied:
Когда методы просто пробрасывают значение то тесты тоже выходят деградированные :)
Сделал verify() на количество вызываний методов репозитория.

HalinaPichukha replied:
Простое пробрасывае тоже покрывается тестами на прод проектах.
+ any() по возможности лучше не использовать.

RomanZubariev replied:
Убрал any(), проверил все методы.

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+
+  @BeforeEach
+  public void setup() {
+    employeeServiceImpl = new EmployeeServiceImpl(employeeRepository);
```
Не разобрался до конца с @InjectMocks раз создаёшь объект EmployeeServiceImpl через new

RomanZubariev replied:
Первоначально это был конструктор и инициализация была там. Забыл убрать.

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
+    Assertions.assertThrows(IllegalArgumentException.class,
+        () -> employeeServiceImpl.save(testEmployee));
+  }
 }
```
В целом лучше, чем в первый раз :) Ну и можно покрыть тестами все методы в сервисе.

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,61 @@
 package com.mastery.java.task.service;
 
+import static org.mockito.Mockito.when;
```
Обычно статические импорты идут после обычных, а импорты можно группировать по пакетам типа import org.junit.jupiter.api.*. Обычно решается на проекте сколько должно быть импортов, чтобы объединить их под *.

RomanZubariev replied:
Я пользовался гугловским форматтером, там было в точности до наоборот - сначала статические, потом обычные. Поменял в настройках.

HalinaPichukha replied:
Сорри, я думала, что это общая практика)))) Просто в глаза бросается мне) Оставляй любой вариант, который твой форматтер предусматривает.

RomanZubariev1990 committed:
More tests and minor code style changes

RomanZubariev1990 committed:
Added slf4j logging with configurations in logback.xml. Minor tweaks to exception handling

RomanZubariev1990 committed:
Added logging

RomanZubariev1990 committed:
A service layer test added.

RomanZubariev1990 committed:
Added spring boot actuator dependency. Loggers endpoint is exposured and enabled to modify logging level at runtime.

HalinaPichukha commented on src/test/java/com/mastery/java/task/service/EmployeeServiceTest.java:
```diff
@@ -1,5 +1,118 @@
+    when(employeeRepository.findById(testEmployee.getEmployeeId())).thenReturn(
+        Optional.of(testEmployee));
+    when(employeeRepository.save(testEmployee)).thenReturn(testEmployee);
+    Assertions.assertTrue(employeeServiceImpl.update(testEmployee).equals(testEmployee));
```
Почему сразу не использовать assertEquals?

RomanZubariev replied:
Поменял на assertEquals везде где это применимо.

HalinaPichukha commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -1,58 +1,54 @@
-      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Employee not found");
-    }
+    employeeRepository.findById(employee.getEmployeeId()).orElseThrow(
+        () -> new EmptyResultDataAccessException(
```
Получается, что ты возвращаешь значение, но не используешь его. Если подключишь какой-нибудь анализатор кода, то заметишь что-то вроде: The return value of "orElseThrow" must be used.
 У репозитория есть другой метод, который не возвращает объект из базы, а просто прверяет наличие. Я бы использовала existsById, а потом уже бросала ошибку, если false. 
Либо использовать объект, который возвращает findById - сетать ему новые данные, а потом сохранять.

RomanZubariev replied:
Заменил на условный оператор от existsById

HalinaPichukha commented on src/main/java/com/mastery/java/task/service/EmployeeServiceImpl.java:
```diff
@@ -1,58 +1,54 @@
-    return employeeDao.save(employee);
+    Optional.ofNullable(employee.getEmployeeId()).ifPresent(id -> {
+      throw new IllegalArgumentException("Employee id should be null.");
+    });
```
На мой взгляд усложнено) Понятно, что ты разобрался с Optional) но можно сделать простую проверку на null.

RomanZubariev replied:
Заменил на условный оператор на null.

HalinaPichukha commented on src/main/java/com/mastery/java/task/rest/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+@ControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
```
Попробуй поставить себе плагин на Сонар, например. Тогда многие вещи, которые не знаешь или которые сразу не очевидны, сразу бросаются в глаза. 
Мой сонар говорит: Having a variable with the same name in two unrelated classes is fine, but do the same thing within a class hierarchy and you’ll get confusion at best, chaos at worst. 
Я тоже не знала, что ResponseEntityExceptionHandler имеет поле Log logger :)
Обычно логер называют log.

RomanZubariev replied:
Поставил сонар, буду смотреть что он пишет.

HalinaPichukha commented on src/main/java/com/mastery/java/task/rest/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+      EmptyResultDataAccessException exception) {
+    if (Arrays.stream(exception.getStackTrace())
+        .anyMatch(te -> te.getMethodName().matches("deleteById"))) {
+      logger.warn("Nothing was deleted.");
```
Почитай про бест практисы по логированию :)
Там будут пункты типа:
Add Context to Your Logs
Don’t Log Too Much or Too Little
Logging Java Exceptions 
и т.д.

Увидеть в логах "Nothing was deleted." - значит ничего не ивидеть)
exception.getMessage() ты и так увидишь) и без логирования. 

Логирование - это ведение журнала, где ты можешь найти информацию, полезную для статистики, например, или для фикса косяков. Но как разобраться, что пошло не так, если мы не знаем деталей?

+ уровни логирования. Почему важно их разделять? Часто логи пишутся в разные файлы (не будем сейчас о тулах для логирования вспоминать, хотя, для них, наверное, ещё и важнее) по уровням логирования, для того, чтобы мониторить ситуацию на проде: начали логироваться errors - поднимут крик. Другие файлы мониторят реже. Поэтому важно понимать, что на каком уровне логировать.

RomanZubariev replied:
Я сделал логирование в два файла - info и warn. Стоит для каждого уровня свой файл делать?
Сам уровень выбирал исходя из rest - если возвращает код 4xx значит error. Не знаю оверкилл ли это. Warn в непонятных ситуациях вроде DELETE несуществующей записи (по rest это OK).
Постараюсь более информативные сообщения писать.

RomanZubariev1990 committed:
Minor changes in service implementation and testing.

- EmployeeServiceImpl.update() doesn't get the employee to be updated first, instead checks implicitly if it exists.
- Minor changes in tests.
- More verbous log message when attempting to delete a  nonexisting record

pavel-tsikhanovich reviewed (commented):
Можешь закрыть все мои комментарии до этого. В целом все выглядит довольно неплохо. Давай на текущий момент сделаем вот что: 

1. Добавь побольше логгирования, в сервисе (или контроллере) например, не только о том, что что-то не так, но и информационных, что происходит и тд и тп. Логги очень полезная информация, если что-то в программе идет не так, у тебя на текущий момент 3 лога на все приложение
2. Конфигурация для логов с помощью xml - ок, у Spring можно это делать легче. Посмотри, пожалуйста как. Можешь не переделывать, но поговорим об этом
3. Имена тестов должны нести смысловую нагрузку, например ты тестируешь положительный результат работы метода - shouldReturnEmployeeById, или же если тестируешь отрицательный кейс - shouldThrowEmptyResultDataAccessExceptionIfEmployeeWithGivenIdDoesNotExist. Можно не прямо так, но хотя бы что-нибудь похожее было бы замечательно. Мы действительно уделяем этому внимание)

RomanZubariev1990 committed:
Added logging to controller and modified tests names.

- Controller methods now log to info level to trace the http requests.
- Test names are now more informative.
